### PR TITLE
Add the shadowing scan activity for shadowing

### PR DESCRIPTION
**Pull request by yux0**
Add the shadowing scan activity for shadowing.

TODO: add unit tests

### CODE REVIEW

File: src/main/java/com/uber/cadence/internal/metrics/MetricsType.java [hunk 1]
```diff
@@ -161,4 +161,11 @@
 
   public static final String NON_DETERMINISTIC_ERROR =
       CADENCE_METRICS_PREFIX + "non-deterministic-error";
+
+  public static final String SHADOWING_NON_DETERMINISTIC_ERROR =
+          CADENCE_METRICS_PREFIX + "shadowing-non-deterministic-error";
+  public static final String SHADOWING_SKIPPED_COUNT =
+          CADENCE_METRICS_PREFIX + "shadowing-skipped-workflow";
+  public static final String SHADOWING_SUCCESS_COUNT =
+          CADENCE_METRICS_PREFIX + "shadowing-success-workflow";
```
Comment (yycptt) [inline src/main/java/com/uber/cadence/internal/metrics/MetricsType.java:170]: shall we make these metrics the same as those defined in go-client?

Also looks like this part should belong to the replay activity diff?

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityRequest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+package com.uber.cadence.internal.shadowing;
+
+public class ScanWorkflowActivityRequest {
+  private final String domain;
+  private final String workflowQuery;
+  private final byte[] pageToken;
+  private final int pageSize;
```
Comment (yycptt) [inline src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityRequest.java:22]: I think we also need to perform sampling in this activity, so probably add `samplingRate` here?

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityResponse.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+package com.uber.cadence.internal.shadowing;
+
+import com.uber.cadence.WorkflowExecution;
+import java.util.List;
+
+public class ScanWorkflowActivityResponse {
+  private final List<WorkflowExecution> workflowExecutions;
```
Comment (yycptt) [inline src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityResponse.java:22]: just want to confirm, is the `WorkflowExecution` defined by thrift or a custom type defined in java client?

File: src/main/java/com/uber/cadence/internal/metrics/MetricsType.java (hunk 1 above)
Comment (yux0) [reply src/main/java/com/uber/cadence/internal/metrics/MetricsType.java]: Moved to replay activity PR

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityRequest.java (hunk 2 above)
Comment (yux0) [reply src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityRequest.java]: Added

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityResponse.java (hunk 3 above)
Comment (yux0) [reply src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityResponse.java]: Adopt your IDL changes

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+              + "; query: "
+              + request.getQuery(),
+          t);
+      throw t;
+    }
+  }
+
+  public List<WorkflowExecution> samplingWorkflows(
+      List<WorkflowExecutionInfo> executionInfoList, double samplingRate) {
+    int capacity = executionInfoList.size();
```
Comment (yycptt) [inline src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java:80]: not related to your diff: I am thinking if after sampling there's only a very small number of executions or no executions at all, we should call scan again inside this activity.

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  public ListWorkflowExecutionsResponse scanWorkflows(ListWorkflowExecutionsRequest request)
+      throws Throwable {
+    try {
+      return RpcRetryer.retryWithResult(
+          RpcRetryer.DEFAULT_RPC_RETRY_OPTIONS,
+          () -> this.serviceClient.ScanWorkflowExecutions(request));
+    } catch (Throwable t) {
+      // TODO: handle non retryable error
```
Comment (yycptt) [inline src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java:67]: do you plan to do this in a separate diff?

Comment (yux0) [reply src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java:67]: Yes.

**Review by yycptt — APPROVED** [review]

File: src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java (hunk 4 above)
Comment (yux0) [reply src/main/java/com/uber/cadence/internal/shadowing/ScanWorkflowActivityImpl.java]: I updated it to at least return 1 execution.

**Pull request merged**